benchmarks/entropy_adiabat/plugins/entropy_advection.cc

```diff
@@ -75,75 +79,94 @@ namespace aspect
-              *
-              (rho_T);
-
-          Tensor<1,dim> current_u = scratch.current_velocity_values[q];
-          // Subtract off the mesh velocity for ALE corrections if necessary
-          if (this->get_parameters().mesh_deformation_enabled)
-            current_u -= scratch.mesh_velocity_values[q];
-
           const double JxW = scratch.finite_element_values.JxW(q);
 
-          // For the diffusion constant, use the larger of the physical
-          // and the artificial viscosity/conductivity/diffusion constant.
-          // One could also choose the sum of the two, but if the
-          // physical diffusion is larger than the artificial one,
-          // then (because the latter is chosen sufficiently large to
-          // make the problem stable) one may as well stick with the
-          // physical one. And if the physical diffusion is too small to
-          // make the problem stable, then we ought to choose the smallest
-          // diffusivity value that makes the problem stable -- which is
-          // exactly the artificial viscosity.
-          const double entropy_conductivity =  scratch.material_model_outputs.thermal_conductivities[q] * scratch.material_model_inputs.temperature[q] /
-                                               scratch.material_model_outputs.specific_heat[q];
-          const double diffusion_constant = std::max (entropy_conductivity,
-                                                      scratch.artificial_viscosity);
-
-          // do the actual assembly. note that we only need to loop over the advection
-          // shape functions because these are the only contributions we compute here
-          for (unsigned int i=0; i<advection_dofs_per_cell; ++i)
+          // solve the diffusion equation for the temperature
+          if (advection_field.is_temperature())
+            {
+              const double density_c_P = scratch.material_model_outputs.densities[q] *
+                                         scratch.material_model_outputs.specific_heat[q];
+
+              const double field_term_for_rhs = scratch.old_field_values[q] * density_c_P;
+
+              // do the actual assembly. note that we only need to loop over the advection
+              // shape functions because these are the only contributions we compute here
+              for (unsigned int i=0; i<advection_dofs_per_cell; ++i)
+                {
+                  data.local_rhs(i)
+                  += field_term_for_rhs * scratch.phi_field[i] * JxW;
+
+                  for (unsigned int j=0; j<advection_dofs_per_cell; ++j)
+                    {
+                      data.local_matrix(i,j)
+                      += (
+                           (time_step * scratch.material_model_outputs.thermal_conductivities[q]
+                            * (scratch.grad_phi_field[i] * scratch.grad_phi_field[j]))
+                           + (scratch.phi_field[i] * scratch.phi_field[j]) * density_c_P
```

Comment:
if BDF2 is possible, we would need to multiply this line with `bdf2_factor` like in the assembly below (line 165).